src/main/scala/workflow/graph/NodeOptimizationRule.scala

```diff
@@ -0,0 +1,200 @@
+package workflow.graph
+
+import org.apache.spark.rdd.RDD
+import workflow.WorkflowUtils
+
+/**
+ * A modified graph executor used by the NodeOptimizationRule, to keep track
```

Comment:
I'm a little confused why this is a separate class and why it doesn't extend anything?

Can you describe what this thing is supposed to actually do? Is it for walking through the graph and collecting samples?